Fix bypassing vaulting clumsy check with verb action.

Tayrtahn opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
Stops clumsy characters being able to climb/vault onto tables and such by using alt-click or the verb menu instead of drag-dropping.

This also fixes a separate bug where, when a clumsy character managed to pass the bonk check, the action would just cancel and the character didn't actually climb the object. Passing the check will now let the character climb.

Additionally, this fixes a oddity where the ability for mobs to climb was determined by whether or not they had feet/legs, which was arbitrary and problematic for creatures like slimes (which should be able to ooze up onto a table). This check could also be bypassed by using the verb instead of drag-dropping, so that was fixed too. The ability for mobs to climb is now controlled in the prototypes.

The default chance of a clumsy character bonking on a climbable has also been reduced from 75% to 50% to compensate somewhat.

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
Fixes #24951
Fixes #17423
Fixes #25951

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
ClimbSystem now raises AttemptClimbEvent as part of TryClimb, and aborts climbing if it gets cancelled. This makes sure that all code paths go through a clumsy check. BonkSystem now listens for AttemptClimbEvents and responds to them instead of trying to intercept DragDropEvents.
CanVault now gets checked in TryClimb as well, to prevent bypassing it.

The logic for climbing capabilities is now:
- The presence of ClimbingComponent makes an entity able to be placed on surfaces like tables.
- ClimbingComponent.CanClimb controls whether it can climb onto surfaces by drag-drop or verb.

The new field defaults to true to minimize changes to existing behavior. Some mobs will have gained the ability to climb when previously they couldn't, but that should be less impactful than the opposite and can be resolved by YML changes in the future.

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->


**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl:
- fix: Clumsy characters can no longer avoid bonking their heads by using the Climb verb.
- tweak: Clumsy characters are less likely to bonk their heads when trying to climb.

themias commented:
I'd imagine this would also fix #17423.

Tayrtahn commented:
Yeah, looks like it does!

EmoGarbage404 commented:
i would rather the bonkable system itself just get fixed considering that it's not very large and this hack is quite nasty.

Tayrtahn commented:
Yeah, fair enough. Maybe I'll take a crack at that soon

themias commented:
To prevent this from being debilitating, we could reduce BonkClumsyChance. (It's currently at 75% !) Dropping it down to, say, 5% would allow clowns to climb most of the time, with the occasional funny bonk.

Tayrtahn commented:
Yeah, I'm not fond of relying on bugs for balance. Better to fix the bug and implement the desired balance properly.

Tayrtahn commented:
It works certainly be possible to make ClumsyComponent have its own likelihood modifier too. So clowns could have a small chance of bonking while monkeys are more likely to bonk. Could even apply a boost when the character is drunk...

Tayrtahn commented:
Lowered the default bonk chance to 50% and made the chance configurable as part of ClumsyComponent, so different entities can be given different levels of clumsiness.

metalgearsloth commented on Content.Shared/Interaction/Components/ClumsyComponent.cs:
```diff
@@ -9,14 +9,21 @@ namespace Content.Shared.Interaction.Components
+        [DataField(required: true)]
         [ViewVariables(VVAccess.ReadWrite)]
         public DamageSpecifier ClumsyDamage = default!;
 
+        /// <summary>
+        /// Chance between 0 and 1 that the character will bonk when
+        /// trying to climb something.
+        /// </summary>
+        [DataField]
+        public float BonkChance = 0.5f;
+
```
Moving 1 bonk field out is pretty weird, would be better to keep it all on the same component.

Tayrtahn replied:
I can put it back, I suppose. Would it be alright to add a ClumsyLevel property to ClumsyComponent and like multiply ClumsyComponent.ClumsyLevel * BonkableComponent.BonkClumsyChance to input into the clumsy roll instead?

I like the idea of clumsiness not being a hard binary, so some creatures can be just a little bit clumsy.

Tayrtahn replied:
On second thought, I'll save that for another PR.

deltanedas commented:
> apply a boost when character is drunk

give every climbing mob clumsy set to 0% base chance then when they get drunk increase so disgruntled bar patrons trying to jump the counter slam their face :trollface:

Tayrtahn commented:
Added a fix for CanVault only being checked with drag-drop interaction. Now the verb can't be used to bypass it.

Also removed the weird check for a leg or foot as part of CanVault. CanVault now checks if the user's ClimbingComponent has CanClimb set to true, or CanForceClimb set to true if dragging another entity onto something.

As discussed in #25951, this may make some things able to climb that couldn't before. This shouldn't be too big of a problem and can be fixed in a future PR.

Tayrtahn commented:
No need to apologize! This is what PR reviews are for, after all.

I think I'll take the simpler route and get rid of CanClimbForce for now.

metalgearsloth commented on Content.Shared/Climbing/Events/AttemptClimbEvent.cs:
```diff
@@ -0,0 +1,15 @@
+namespace Content.Shared.Climbing.Events;
+
+public sealed class AttemptClimbEvent : CancellableEntityEventArgs
+{
+    public EntityUid User;
+    public EntityUid Climber;
+    public EntityUid Climbable;
+
+    public AttemptClimbEvent(EntityUid user, EntityUid climber, EntityUid climbable)
+    {
+        User = user;
+        Climber = climber;
+        Climbable = climbable;
+    }
+}
```
record struct byref plz, rest looks good.
